File: src/main/java/hello/torrent/YTS.java
package hello.torrent;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import hello.utilities.HttpHelper;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicReference;

public class YTS extends HttpUser implements TorrentSearchEngine {

    YTS(HttpHelper httpHelper) {
        super(httpHelper);
    }

    @Override
    public List<Torrent> searchTorrents(String searchName) {

        CopyOnWriteArrayList<Torrent> torrentList = new CopyOnWriteArrayList<>();

        String resultString = null;
        resultString = httpHelper.getPage(buildSearchUrl(searchName));

        torrentList.addAll(parseTorrentsOnResultPage(resultString, searchName));
        torrentList.sort(TorrentHelper.torrentSorter);
        return torrentList;
    }

    private String buildSearchUrl(String searchName) {
        return String.format(getBaseUrl() + "/api/v2/list_movies.json?limit=50&query_term=%s&sort_by=seeds", URLEncoder.encode(searchName, StandardCharsets.UTF_8));
    }

    @Override
    public String getBaseUrl() {
        return "https://yts.mx";
    }

    @Override
    public String getSearchPage() {
        return buildSearchUrl("test");
    }

    private List<Torrent> parseTorrentsOnResultPage(String pageContent, String searchName) {
        ArrayList<Torrent> torrentList = new ArrayList<>();

        JsonParser parser = new JsonParser();
        JsonElement jsonRoot = parser.parse(pageContent);
        JsonElement data = jsonRoot.getAsJsonObject().get("data");
        JsonElement results = data.getAsJsonObject().get("movies");
        JsonArray jsonArray = results.getAsJsonArray();
        jsonArray.forEach(jsonTorrentElement -> {
            Torrent tempTorrent = new Torrent(toString());
            final JsonObject jsonTorrent = jsonTorrentElement.getAsJsonObject();
            tempTorrent.name = jsonTorrent.get("title").getAsString() + " " + jsonTorrent.get("year").getAsInt();
            JsonObject bestTorrentSource = retrieveBestTorrent(jsonTorrent.get("torrents").getAsJsonArray());

            //tempTorrent.magnetUri = bestTorrentSource.get("url").getAsString();
            tempTorrent.magnetUri = TorrentHelper.buildMagnetUriFromHash(bestTorrentSource.get("hash").getAsString().toLowerCase(), tempTorrent.name);
            tempTorrent.seeder = bestTorrentSource.get("seeds").getAsInt();
            tempTorrent.leecher = bestTorrentSource.get("peers").getAsInt();
            tempTorrent.size = bestTorrentSource.get("size").getAsString();
            tempTorrent.lsize = bestTorrentSource.get("size_bytes").getAsLong() / 1024.0f / 1024.0f;
            tempTorrent.date = new Date(bestTorrentSource.get("date_uploaded_unix").getAsLong() * 1000);

            TorrentHelper.evaluateRating(tempTorrent, searchName);
            if (TorrentHelper.isValidTorrent(tempTorrent)) {
                torrentList.add(tempTorrent);
            }
        });

        return torrentList;
    }

    private JsonObject retrieveBestTorrent(JsonArray torrentElements) {
        AtomicReference<JsonObject> bestTorrent = new AtomicReference<>();
        torrentElements.forEach(torrentElement -> {
            if (bestTorrent.get() == null || bestTorrent.get().get("size_bytes").getAsLong() < torrentElement.getAsJsonObject().get("size_bytes").getAsLong()) {
                bestTorrent.set(torrentElement.getAsJsonObject());
            }
        });
        return bestTorrent.get();
    }

    @Override
    public Torrent suggestATorrent(List<Torrent> inputList) {
        return inputList.stream().max(TorrentHelper.torrentSorter).orElse(null);
    }

    @Override
    public String toString() {
        return this.getClass().getSimpleName();
    }

}

File: src/main/java/boat/torrent/searchEngines/YTS.kt
package boat.torrent.searchEngines

import boat.torrent.HttpUser
import boat.torrent.Torrent
import boat.torrent.TorrentComparator
import boat.torrent.TorrentHelper
import boat.torrent.TorrentSearchEngine
import boat.utilities.HttpHelper
import boat.utilities.LoggerDelegate
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicReference
import java.util.function.Consumer

class YTS internal constructor(httpHelper: HttpHelper) : HttpUser(httpHelper), TorrentSearchEngine {

    companion object {
        private val logger by LoggerDelegate()
    }

    override fun searchTorrents(searchName: String): List<Torrent> {
        val torrentList = CopyOnWriteArrayList<Torrent>()
        val resultString = httpHelper.getPage(buildSearchUrl(searchName))
        torrentList.addAll(parseTorrentsOnResultPage(resultString, searchName))
        torrentList.sortWith(TorrentComparator)
        return torrentList
    }

    private fun buildSearchUrl(searchName: String): String {
        return String.format(
                "%s/api/v2/list_movies.json?limit=50&query_term=%s&sort_by=seeds",
                baseUrl,
                URLEncoder.encode(searchName, StandardCharsets.UTF_8)
        )
    }

    override fun getBaseUrl(): String {
        return "https://yts.mx"
    }

    private fun parseTorrentsOnResultPage(pageContent: String, searchName: String): List<Torrent> {
        val torrentList = ArrayList<Torrent>()
        try {
            val jsonRoot = JsonParser.parseString(pageContent)
            if (jsonRoot == null || !jsonRoot.isJsonObject) {
                return torrentList
            }
            val data = jsonRoot.asJsonObject["data"] ?: return torrentList
            val results = data.asJsonObject["movies"] ?: return torrentList
            val jsonArray = results.asJsonArray
            jsonArray.forEach(Consumer { jsonTorrentElement: JsonElement ->
                val tempTorrent = Torrent(toString())
                val jsonTorrent = jsonTorrentElement.asJsonObject
                tempTorrent.name = jsonTorrent["title"].asString + " " + jsonTorrent["year"].asInt
                val bestTorrentSource = retrieveBestTorrent(jsonTorrent["torrents"].asJsonArray)
                tempTorrent.isVerified = true
                tempTorrent.magnetUri = TorrentHelper.buildMagnetUriFromHash(bestTorrentSource!!["hash"].asString.lowercase(Locale.getDefault()), tempTorrent.name)
                tempTorrent.seeder = bestTorrentSource["seeds"].asInt
                tempTorrent.leecher = bestTorrentSource["peers"].asInt
                tempTorrent.size = bestTorrentSource["size"].asString
                tempTorrent.sizeInMB = (bestTorrentSource["size_bytes"].asLong / 1024.0f / 1024.0f).toDouble()
                tempTorrent.date = Date(bestTorrentSource["date_uploaded_unix"].asLong * 1000)
                TorrentHelper.evaluateRating(tempTorrent, searchName)
                if (TorrentHelper.isValidTorrent(tempTorrent)) {
                    torrentList.add(tempTorrent)
                }
            })
        } catch (exception: Exception) {
            logger.error("failed to parse string:\n${pageContent}", exception)
        }
        return torrentList
    }

    private fun retrieveBestTorrent(torrentElements: JsonArray): JsonObject? {
        val bestTorrent = AtomicReference<JsonObject?>()
        torrentElements.forEach(Consumer { torrentElement: JsonElement ->
            if (bestTorrent.get() == null || bestTorrent.get()!!["size_bytes"].asLong < torrentElement
                            .asJsonObject["size_bytes"].asLong
            ) {
                bestTorrent.set(torrentElement.asJsonObject)
            }
        })
        return bestTorrent.get()
    }

    override fun toString(): String {
        return this.javaClass.simpleName
    }
}